README.md
# vulfocus-plus 
- forks from [vulfocus](https://github.com/fofapro/vulfocus)

## 全部使用docker安装
> 启动 `docker`
```bash
#!/bin/bash
rm -rf /var/run/docker*

# mount
mount -t nfs 10.27.106.174:/home/nfs /mnt/docker/ -o proto=tcp -o nolock

## docker
/usr/bin/dockerd --registry-mirror=https://hslzset5.mirror.aliyuncs.com -H unix:///var/run/docker.sock  -H tcp://0.0.0.0:2375 --data-root=/mnt/docker
```

## 安装
```bash
docker-compose up -d 

# 创建`mysql`数据库
docker exec -t mysql mysql -uroot -ptest@1q2w2e4R -e 'drop database vulcofus;'
docker exec -t mysql mysql -uroot -ptest@1q2w2e4R -e 'CREATE DATABASE `vulcofus` DEFAULT CHARACTER SET utf8 COLLATE utf8_general_ci;'

```



CHANGELOG.md
## 更新日志

### v0.4.0.1

更新时间 2021/1/25 

- 部署模式更新，增加 `docker-compose` 以及平台部署组件的单镜像管理
- 修改`sqlite3`为`mysql`
- 修改从环境变量加载配置为从`config.yaml`加载
- 修改 `django`,`docker-compose` 的版本为最新

### v0.3.2.1

更新时间：2020-12-25

- 新增多环境编排功能
- 新增网卡管理功能
- 新增可视化配置编排功能
- 新增场景模式
- 新增10个漏洞镜像
- 迁移管理中心、镜像管理、靶场管理
- 修复漏洞环境删除积分丢失 BUG
- 修复个人主页显示错误 BUG
- 修复一些不为人知的 BUG

### v0.3.1.1

更新时间：2020-09-04

- 新增 8 个镜像
- 新增内置 83 个镜像
- 修复一些 BUG

### v0.2.3

更新时间：2020-07-02

- 新增默认内置镜像 82 个镜像
- 新增漏洞环境查询功能
- 新增镜像信息修改功能
- 优化镜像删除时提示功能

### v0.2.2.1

更新时间：2020-06-13

- 修复 [@UUUUnotfound](https://github.com/UUUUnotfound) 提出的 `SECRET_KEY` 泄漏安全漏洞

### v0.2.2

更新时间：2020-05-29

- 新增翻页功能
- 新增进度条，可展示目前 `pull`、`push` 镜像的进度信息
- 新增镜像分享功能
- 修改一些已知 BUG

###  v0.2.1

更新时间：2020-05-08

- 优化前端显示漏洞镜像错位问题
- 优化容器到期时间

### v0.2.0

更新时间：2020-05-08

- 新增定时功能，可自定义关闭漏洞环境
- 新增本地环境加载功能，可一键导入本地所有漏洞环境
- 新增镜像搜索方式
- 优化漏洞镜像下载流程
- 优化漏洞提交设置默认 rank 值
- 优化 Docker 连接方式，新增 IP 连接方式
- 兼容 Windows 环境 Docker



vulfocus-api/vulfocus/urls.py
"""vulbox URL Configuration

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/2.2/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  path('', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.urls import include, path
    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
"""
from django.conf.urls import url, include
from rest_framework import routers
from dockerapi.views import ImageInfoViewSet, ContainerVulViewSet, SysLogSet, get_setting, update_setting
from user.views import UserRegView, UserSet
from rest_framework_jwt.views import obtain_jwt_token
from user.views import get_user_info, LogoutView
from tasks.views import TaskSet
from network.views import NetWorkInfoViewSet
from layout_image.views import LayoutViewSet, upload_img

router = routers.DefaultRouter()
router.register('images', ImageInfoViewSet, basename='Images')
router.register('container', ContainerVulViewSet, basename='Container')
router.register('user/register', UserRegView, basename='register')
router.register('user', UserSet, basename='user')
router.register('syslog', SysLogSet, basename="SysLog")
router.register('tasks', TaskSet, basename="TaskSet")
router.register("network", NetWorkInfoViewSet, basename="network")
router.register('layout', LayoutViewSet, basename="layout")

urlpatterns = [
    url(r'^', include(router.urls)),
    url(r'^user/login', obtain_jwt_token),
    url(r'^user/logout', LogoutView.as_view(), name="logout"),
    url(r'user/info', get_user_info.as_view()),
    url(r'setting/get', get_setting),
    url(r'setting/update', update_setting),
    url(r'img/upload', upload_img),
    url(r'^api-auth/', include('rest_framework.urls', namespace='rest_framework')),
]


vulfocus-api/layout_image/bridge.py
"""
bridge to docker-compose
"""

import logging
from os.path import normpath
from compose import __version__ as compose_version
from compose.container import Container
from compose.cli.utils import get_version_info
from compose.cli.command import get_project as compose_get_project, get_config_path_from_options, get_config_from_options
from compose.config.config import get_default_config_files
from compose.config.environment import Environment

from compose.cli.docker_client import docker_client
from compose.const import API_VERSIONS, ComposeVersion
COMPOSEFILE_V3_0 = ComposeVersion('3.0')

logging.info(get_version_info('full'))


def ps_(project):
    """
    containers status
    """
    logging.info('ps ' + project.name)
    running_containers = project.containers(stopped=True)

    items = [{
        'name': container.name,
        'name_without_project': container.name_without_project,
        'command': container.human_readable_command,
        'state': container.human_readable_state,
        'labels': container.labels,
        'ports': container.ports,
        'volumes': get_volumes(get_container_from_id(project.client, container.id)),
        'is_running': container.is_running} for container in running_containers]

    return items


def get_container_from_id(my_docker_client, container_id):
    """
    return the docker container from a given id
    """
    return Container.from_id(my_docker_client, container_id)


def get_volumes(container):
    """
    retrieve container volumes details
    """
    mounts = container.get('Mounts')
    return [dict(source=mount['Source'], destination=mount['Destination']) for mount in mounts]


def get_yml_path(path):
    """
    get path of docker-compose.yml file
    """
    return get_default_config_files(path)[0]


def get_project(path):
    """
    get docker project given file path
    """
    logging.debug('get project ' + path)

    environment = Environment.from_env_file(path)
    config_path = get_config_path_from_options(path, dict(), environment)
    project = compose_get_project(path, config_path)
    return project


def containers():
    """
    active containers
    """
    return client().containers()


def info():
    """
    docker info
    """
    docker_info = client().info()
    return dict(compose=compose_version, info=docker_info['ServerVersion'], name=docker_info['Name'])


def client():
    """
    docker client
    """
    return docker_client(Environment(), API_VERSIONS[COMPOSEFILE_V3_0])


def project_config(path):
    """
    docker-compose config
    """
    norm_path = normpath(path)
    return get_config_from_options(norm_path, dict())


vulfocus-api/layout_image/migrations/0014_auto_20210125_1432.py
# Generated by Django 3.1 on 2021-01-25 14:32

from django.db import migrations, models
import uuid


class Migration(migrations.Migration):

    dependencies = [
        ('layout_image', '0013_auto_20201223_2220'),
    ]

    operations = [
        migrations.AlterField(
            model_name='layout',
            name='layout_id',
            field=models.UUIDField(default=uuid.UUID('b42777df-02fc-4a7e-a8c6-dc9d5d847a20'), editable=False, primary_key=True, serialize=False, verbose_name='编排UUID'),
        ),
        migrations.AlterField(
            model_name='layoutdata',
            name='layout_user_id',
            field=models.UUIDField(default=uuid.UUID('56487b3b-0cdd-402a-b9a5-c81453a0ef71'), editable=False, primary_key=True, serialize=False, verbose_name='ID'),
        ),
        migrations.AlterField(
            model_name='layoutservice',
            name='service_id',
            field=models.UUIDField(default=uuid.UUID('4d98667f-3313-4c1a-b233-08f697cdeffe'), editable=False, primary_key=True, serialize=False, verbose_name='ID'),
        ),
        migrations.AlterField(
            model_name='layoutservicecontainer',
            name='service_container_id',
            field=models.UUIDField(default=uuid.UUID('96c081f8-d897-4c3e-8947-bd96d04c07e1'), editable=False, primary_key=True, serialize=False, verbose_name='ID'),
        ),
        migrations.AlterField(
            model_name='layoutservicecontainerscore',
            name='layout_service_container_score_id',
            field=models.UUIDField(default=uuid.UUID('3e0c0c55-09ce-4cff-b303-312b62d7b70b'), editable=False, primary_key=True, serialize=False, verbose_name='ID'),
        ),
        migrations.AlterField(
            model_name='layoutservicenetwork',
            name='layout_service_network_id',
            field=models.UUIDField(default=uuid.UUID('a5d28af1-2dd2-485c-8ebf-f457dc155b02'), editable=False, primary_key=True, serialize=False, verbose_name='ID'),
        ),
    ]


vulfocus-api/vulfocus/settings.py
"""
Django settings for vulfocus project.

Generated by 'django-admin startproject' using Django 2.2.5.

For more information on this file, see
https://docs.djangoproject.com/en/2.2/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/2.2/ref/settings/
"""

import os
import docker
import datetime
import redis
from django.core.management import utils

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# TODO 2021-1-25 load from `config.yaml`
from .conf import config

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/2.2/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = config.SECRET_KEY


# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = config.DEBUG

ALLOWED_HOSTS = ["*"]

AUTH_USER_MODEL = "user.UserProfile"

# Application definition
ALLOWED_IMG_SUFFIX = ["jpg", "jpeg", "png"]

STATICFILES_DIRS = [
    os.path.join(BASE_DIR, "static")
]

INSTALLED_APPS = [
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',

    'rest_framework',
    'user',
    # 'corsheaders',
    'dockerapi',
    'network',
    'tasks',
    'layout_image'
]

# redis host
REDIS_HOST = config.REDIS_HOST
# redis port
REDIS_PORT = config.REDIS_PORT
# redis pass
REDIS_PASS = config.REDIS_PASSWORD
REDIS_DB_CELERY = config.REDIS_DB_CELERY

CELERY_BROKER_URL = f"redis://:{REDIS_PASS}@{REDIS_HOST}:{REDIS_PORT}/{REDIS_DB_CELERY}"
REDIS_POOL = redis.ConnectionPool(host=REDIS_HOST, port=int(REDIS_PORT), password=REDIS_PASS, decode_responses=True, db=1)

#: Only add pickle to this list if your broker is secured
#: from unwanted access (see userguide/security.html)
CELERY_ACCEPT_CONTENT = ['json']
CELERY_RESULT_BACKEND = CELERY_BROKER_URL
CELERY_TASK_SERIALIZER = 'json'

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    # 'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

# CORS_ORIGIN_WHITELIST = [
#     "http://localhost:9527",
# ]
# CORS_ORIGIN_ALLOW_ALL = True

REST_FRAMEWORK = {
    # Use Django's standard `django.contrib.auth` permissions,
    # or allow read-only access for unauthenticated users.
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.IsAuthenticated'
    ],
    'DEFAULT_AUTHENTICATION_CLASSES': [
        'rest_framework_jwt.authentication.JSONWebTokenAuthentication',
        'rest_framework.authentication.SessionAuthentication',
        'rest_framework.authentication.BasicAuthentication',
    ],
    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.PageNumberPagination',
    'PAGE_SIZE': 20
}

JWT_AUTH = {
    # 'JWT_RESPONSE_PAYLOAD_HANDLER':'user.jwt.jwt_response_payload_handler',
    # 指明token的有效期
    'JWT_EXPIRATION_DELTA': datetime.timedelta(days=1),
    'JWT_AUTH_HEADER_PREFIX': 'BMH',

}


ROOT_URLCONF = 'vulfocus.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, 'templates')],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'vulfocus.wsgi.application'


# Database
# https://docs.djangoproject.com/en/2.2/ref/settings/#databases

# Session 和y页面的缓存存储, 由于不是ORM，所以很多用户上

SESSION_COOKIE_DOMAIN = config.SESSION_COOKIE_DOMAIN
CSRF_COOKIE_DOMAIN = config.CSRF_COOKIE_DOMAIN
SESSION_COOKIE_AGE = config.SESSION_COOKIE_AGE
SESSION_EXPIRE_AT_BROWSER_CLOSE = config.SESSION_EXPIRE_AT_BROWSER_CLOSE
SESSION_ENGINE = 'redis_sessions.session'
SESSION_REDIS = {
    'host': config.REDIS_HOST,
    'port': config.REDIS_PORT,
    'password': config.REDIS_PASSWORD,
    'db': config.REDIS_DB_SESSION,
    'prefix': 'auth_session',
    'socket_timeout': 1,
    'retry_on_timeout': False
}

MESSAGE_STORAGE = 'django.contrib.messages.storage.cookie.CookieStorage'
# Database
# https://docs.djangoproject.com/en/1.10/ref/settings/#databases

DB_OPTIONS = {}
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.{}'.format(config.DB_ENGINE.lower()),
        'NAME': config.DB_NAME,
        'HOST': config.DB_HOST,
        'PORT': config.DB_PORT,
        'USER': config.DB_USER,
        'PASSWORD': config.DB_PASSWORD,
        'ATOMIC_REQUESTS': True,
        'OPTIONS': DB_OPTIONS
    },
    # 'mongo': {
    #     'ENGINE': None,
    # }
}

DB_CA_PATH = os.path.join(BASE_DIR, 'data', 'ca.pem')
if config.DB_ENGINE.lower() == 'mysql':
    # DB_OPTIONS['init_command'] = "SET sql_mode='STRICT_TRANS_TABLES'"
    DB_OPTIONS['sql_mode'] = "traditional"
    if os.path.isfile(DB_CA_PATH):
        DB_OPTIONS['ssl'] = {'ca': DB_CA_PATH}
elif config.DB_ENGINE.lower() == 'sqlite3':
    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
        }
    }

# bug MySQL does not allow unique CharFields to have a max_length > 255
SILENCED_SYSTEM_CHECKS = ['mysql.E001']

# Password validation
# https://docs.djangoproject.com/en/2.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/2.2/topics/i18n/
LANGUAGE_CODE = 'zh-hans'

TIME_ZONE = 'Asia/Shanghai'

# TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = False

# 默认启动容器最长时间为 60s，可根据实际情况调整
DOCKER_CONTAINER_TIME = 60

DOCKER_HOST = config.DOCKER_HOST
DOCKER_URL = f"tcp://{DOCKER_HOST}:2375"

if DOCKER_URL.startswith("unix:"):
    client = docker.DockerClient(base_url=DOCKER_URL)
    api_docker_client = docker.APIClient(base_url=DOCKER_URL)
else:
    client = docker.DockerClient(DOCKER_URL)
    api_docker_client = docker.APIClient(base_url=DOCKER_URL)

try:
    """
    设置 docker-compose 连接
    """
    os.environ['DOCKER_HOST'] = DOCKER_URL
except Exception as e:
    pass


# 靶场绑定 IP，提供用户访问靶场与 Docker 服务IP保持一致。
try:
    VUL_IP = config.VUL_IP
except:
    VUL_IP = config.DOCKER_HOST

DOCKER_COMPOSE = os.path.join(BASE_DIR, "docker-compose")

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/2.2/howto/static-files/

STATIC_URL = '/static/'
STATIC_ROOT = os.path.join(BASE_DIR, 'collect_static')
